Reorder and polish spec (#15)
microsoft/did-x509: specification.md

The main difference to other DID methods is that did:x509 requires a certificate chain to be passed using a new [DID resolution option](https://www.w3.org/TR/did-core/#did-resolution-options) `x509chain` while resolving a DID. This certificate chain is typically embedded in the signing envelope, for example within the `x5c` header parameter of JWS/JWT documents.

> Embedding certificate chains in configuration or policy is cumbersome and often too coarse to identify the intended certificate holder.

Reviewer: Too coarse or too unstable if elements in the chain are short-lived.

version            = 1*DIGIT
ca-fingerprint-alg = "sha256" / "sha384" / "sha512"
ca-fingerprint     = base64url

> predicate-name     = "subject" / "san" / "eku" / "fulcio-issuer"

Author: Context: `1*ALPHA` does not allow `fulcio-issuer`

predicate-value    = *(1*idchar ":") 1*idchar
base64url          = 1*(ALPHA / DIGIT / "-" / "_")
```

The current version value is `0`.

The `ca-fingerprint-alg` value is one of `sha256`, `sha384`, or `sha512`. The `ca-fingerprint` value is a base64url-encoded digest of a non-leaf certificate in the certificate chain, that is, either an intermediate or root CA certificate.

> The `::` separator introduces predicates. Each predicate has a `predicate-name` and a predicate-specific `predicate-value`.

Reviewer: ```suggestion
The `::` separator introduces predicates. Each predicate has a `predicate-name` and a `predicate-value`.
```

Lighter and no less clear in my opinion.

## Method-specific Identifier

The method-specific identifier has three parts:

1. A version number.
2. A certificate authority fingerprint algorithm and value.
3. One or more predicates that match fields in the leaf certificate.

did:x509 does not define any DID URL path or query semantics. A did:x509 DID URL MUST NOT include a path or query component. Fragment identifiers remain valid for identifying resources within a resolved DID document, for example `<DID>#0`.

Example:

`did:x509:0:sha256:WE4P5dd8DnLHSkyHaIjhp4udlkF9LqoKwCvu9gl38jk::subject:C:US:ST:California:O:Example%20Organisation`

In this example, the identifier pins to a certificate authority using a SHA-256 certificate hash and uses the `subject` predicate to express criteria that a leaf certificate subject must fulfil. This identifier will match certificate chains with matching leaf certificate subject fields and a matching intermediate or root CA certificate.

### Predicate validation model

> Predicate validation is defined in [Rego](https://www.openpolicyagent.org/docs/latest/policy-language/) to avoid ambiguous pseudo-code. Implementations do not have to use Rego.

Reviewer: ```suggestion
Predicate validation is defined in [Rego](https://www.openpolicyagent.org/docs/latest/policy-language/) to avoid ambiguous pseudo-code. This has no bearing on implementations, which can be written in any language.
```

For the reference Rego, pass the DID string and parsed certificate-chain JSON as:

```json
{
  "did": "<DID>",
  "chain": [
    {
      "fingerprint": {
        "sha256": "<leaf-sha256>"
      },
      "subject": {
        "CN": "Example"
      },
      "extensions": {}
    },
    {
      "fingerprint": {
        "sha256": "<ca-sha256>"
      },
      "subject": {
        "CN": "Example CA"
      },
      "extensions": {}
    }
  ]
}
```

> Here, `chain` is derived from the `x509chain` resolution option.

Reviewer: Do we need to introduce and define that option?

Author: I see it was there before, but as `"chain": <CertificateChain>`. I think it wasn't necessary to expand it, as it repeats the previously defined chain json. I'll revert this.

### Percent-encoding

Some predicates require values to be percent-encoded. Percent-encoding is specified in [RFC 3986 Section 2.1](https://www.rfc-editor.org/rfc/rfc3986#section-2.1). All characters that are not in the allowed set below must be percent-encoded:

```abnf
allowed = ALPHA / DIGIT / "-" / "." / "_"
```

Note that most libraries implement percent-encoding in the context of URLs and do not encode `~` (`%7E`).

### `subject` predicate

```abnf
predicate-name     = "subject"
predicate-value    = key ":" value *(":" key ":" value)
key                = label / oid

> value              = 1*idchar

Author: These are independent blocks, not repeated grammar

        v := urlquery.decode(items[i+1])
    }
    count(subject) >= 1

>     count(subject) == count(items) / 2

Author: Context: Rego didn't disallow repeated values, when the spec claims they aren't allowed

    object.subset(input.chain[0].subject, subject) == true
}
```

### `san` predicate

```abnf
predicate-name     = "san"
predicate-value    = san-type ":" san-value
san-type           = "email" / "dns" / "uri"

> san-value          = 1*idchar

Author: These are independent blocks, not repeated grammar

`did:x509:0:sha256:WE4P5dd8DnLHSkyHaIjhp4udlkF9LqoKwCvu9gl38jk::eku:1.3.6.1.4.1.311.10.3.13`

Rego policy:

```rego
validate_predicate(name, value) := true if {
    name == "eku"
    value == input.chain[0].extensions.eku[_]
}
```

### `fulcio-issuer` predicate

```abnf
predicate-name     = "fulcio-issuer"
predicate-value    = fulcio-issuer

> fulcio-issuer      = 1*idchar

Author: It's a literal value, not grammar?

}
```

### `eku` predicate

```abnf
predicate-name     = "eku"
predicate-value    = eku
eku                = oid

> oid                = 1*DIGIT *("." 1*DIGIT)

Author: These are independent blocks, not repeated grammar

`did:x509:0:sha256:WE4P5dd8DnLHSkyHaIjhp4udlkF9LqoKwCvu9gl38jk::eku:1.3.6.1.4.1.311.10.3.13`

Rego policy:

```rego
validate_predicate(name, value) := true if {
    name == "eku"
    value == input.chain[0].extensions.eku[_]
}
```

### `fulcio-issuer` predicate

```abnf
predicate-name     = "fulcio-issuer"
predicate-value    = fulcio-issuer

> fulcio-issuer      = 1*idchar

Author: These are independent blocks, not repeated grammar

}
```

## Verifiable Data Registry and Trust Model

did:x509 does not define a persistent registry of DID Documents. Resolution uses the DID string and the `x509chain` resolution option.

The `x509chain` option carries the certificate chain as a comma-separated list of base64url-encoded DER certificates:

```text
x509chain = b64url(DER(leaf)) "," b64url(DER(intermediate)) "," b64url(DER(root))
```

The chain is ordered leaf first and root or trust anchor last. Each comma-separated item is the DER encoding of one complete X.509 certificate, not a public key, fingerprint, or DER encoding of the whole chain.

Trust is established by validating the certificate chain, matching the CA fingerprint, and validating the predicates against the leaf certificate. Applications can add revocation, certificate transparency, signing time, or endorsement checks.

## Certificate Chain JSON Model

> For predicate evaluation, the resolver maps the certificate chain to a small JSON model. This model contains only the fields did:x509 matches on; it does not replace X.509 parsing or RFC 5280 path validation.

Reviewer: ```suggestion
For predicate evaluation, the resolver maps the certificate chain to a limited JSON model. This model contains only the fields did:x509 matches on; it does not replace X.509 parsing or RFC 5280 path validation.
```

As an example, a certificate authority may use email addresses as a way to separate identities and use the SAN extension to store the email address. In that case, the did:x509 identifier should be constructed using the `san` predicate, for example, `did:x509:0:sha256:<ca-fingerprint>::san:email:bob%40example.com`.

In other cases, an authority may not include email addresses at all and instead rely on a specific set of subject fields to separate identities. In that case, the `subject` predicate should be used.

In yet other cases, authorities may assign unique numbers or other types of stable identifiers to logical identities. Typically, this is done to have a stable reference even if a person changes their name or email address.

In all cases, the goal is to craft a did:x509 identifier that is stable yet not too loose in its predicates. An example of a loose did:x509 identifier may be to use the `subject` predicate and only include the `O` field without location fields like country (`C`) or state/locality (`ST`).

> Whether a did:x509 identifier should pin to an intermediate CA instead of a root CA depends on whether there is value in distinguishing between them. Pinning to an intermediate CA typically means that the lifetime of the did:x509 identifier will be shorter, since intermediate CA certificates typically have a shorter validity period than root CA certificates.

Reviewer: ```suggestion
Whether a did:x509 identifier should pin to an intermediate CA instead of a root CA depends on whether there is value in distinguishing between them. Pinning to an intermediate CA typically means that the lifetime of the did:x509 identifier will be shorter, since intermediate CA certificates usually have a shorter validity period than root CA certificates.
```

The `x509chain` resolution option carries the certificate chain used as resolution evidence. Certificates can contain additional metadata beyond the predicates encoded in the DID, including subject attributes, SAN entries, validity periods, certificate policies, and extension values. Resolvers and verifiers should treat certificate chains as potentially identifying data, avoid unnecessary logging or redistribution, and apply data minimization when retaining resolution inputs or outputs.

Stable did:x509 identifiers can enable correlation across transactions, transparency logs, ledgers, and verifiable credentials. If unlinkability is required, relying parties should avoid reusing the same did:x509 identifier across contexts, and issuers should prefer predicates based on role- or service-specific identifiers rather than human-identifying certificate fields.

> ## Examples and Test Vectors

Reviewer: Should we include the test certs inline as base64, or at least a URL to a stable location where they can be found (e.g. tagged commit in this repo)?
